File: src/main/java/com/github/svegon/capi/event/entity/EntityStatusEffectCallback.java
package com.github.svegon.capi.event.entity;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffect;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface EntityStatusEffectCallback {
    Event<EntityStatusEffectCallback> EVENT = EventFactory.createArrayBacked(EntityStatusEffectCallback.class,
            (entity, effect, cir) -> {}, listeners -> (entity, effect, cir) -> {
        for (EntityStatusEffectCallback listener : listeners) {
            listener.hasStatusEffect(entity, effect, cir);

            if (cir.isCancelled()) {
                return;
            }
        }
    });

    void hasStatusEffect(LivingEntity entity, StatusEffect effect, CallbackInfoReturnable<Boolean> cir);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/entity/EntityStatusEffectCallback.kt
package io.github.svegon.mclientapi.event.entity

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.core.Holder
import net.minecraft.world.effect.MobEffect
import net.minecraft.world.entity.LivingEntity
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

fun interface EntityStatusEffectCallback {
    fun hasStatusEffect(entity: LivingEntity, effect: Holder<MobEffect>,
                        cir: CallbackInfoReturnable<Boolean>)

    companion object {
        val EVENT: Event<EntityStatusEffectCallback> = EventFactory.createArrayBacked(
            EntityStatusEffectCallback::class.java,
            EntityStatusEffectCallback { entity, effect, cir -> }) { listeners ->
            EntityStatusEffectCallback { entity, effect, cir ->
                for (listener in listeners) {
                    listener.hasStatusEffect(entity, effect, cir)

                    if (cir.isCancelled) {
                        return@EntityStatusEffectCallback
                    }
                }
            }
        }
    }
}
